File: 01.Welcome to Design Patterns: intro to Design Patterns/Go/In-memory-cache/cache.go
package main

type cache struct {
	storage map[interface{}]interface{}
	evictionAlgo ICache
	capacity int
	maxCapacity int
}

func initCache(e ICache) *cache{
	storage := make(map[interface{}]interface{})
	return &cache{
		storage: storage,
		evictionAlgo: e,
		capacity: 0,
		maxCapacity: 10,

	}
}

func (c *cache) setEvictionAlgo(e ICache) {
    c.evictionAlgo = e
}

func (c *cache) add(key, value string) {
    if c.capacity == c.maxCapacity {
        c.evict()
    }
    c.capacity++
    c.storage[key] = value
}

func (c *cache) get(key string) {
    delete(c.storage, key)
}

func (c *cache) evict() {
    c.evictionAlgo.evict(c)
    c.capacity--
}

File: 01.Welcome to Design Patterns: intro to Design Patterns/Python/In-memory-cache/cache.py
from icache import CacheStrategy

class Cache():
    def __init__(self, cache_strategy : CacheStrategy) -> None:
        self.cache_strategy = cache_strategy
    
    def get_cache_strategy(self) -> CacheStrategy:
        return self.cache_strategy
    
    def set_strategy(self, cache_strategy: CacheStrategy) -> None:
        self.cache_strategy = cache_strategy
    

    def do_some_business_logic(self) -> None:
        self.cache_strategy.evict()